enclave/src/rpc/rpc_get_balance.rs

```diff
@@ -0,0 +1,162 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+pub extern crate alloc;
+use alloc::{boxed::Box, string::String, string::ToString};
+
+use crate::polkadex_balance_storage::Balances;
+use crate::rpc::polkadex_rpc_gateway::RpcGateway;
+use crate::rpc::rpc_call_encoder::{JsonRpcCallEncoder, RpcCall, RpcCallEncoder};
+use crate::rpc::rpc_info::RpcCallStatus;
+use crate::rpc::trusted_operation_verifier::TrustedOperationExtractor;
+use jsonrpc_core::{BoxFuture, Params, Result as RpcResult, RpcMethodSync, Value};
+use log::*;
+use polkadex_sgx_primitives::types::DirectRequest;
+use substratee_stf::{Getter, TrustedGetter, TrustedOperation};
+use substratee_worker_primitives::DirectRequestStatus;
+
+pub struct RpcGetBalance {
+    top_extractor: Box<dyn TrustedOperationExtractor + 'static>,
+    rpc_gateway: Box<dyn RpcGateway + 'static>,
+}
+
+impl RpcGetBalance {
+    pub fn new(
+        top_extractor: Box<dyn TrustedOperationExtractor + 'static>,
+        rpc_gateway: Box<dyn RpcGateway + 'static>,
+    ) -> Self {
+        RpcGetBalance {
+            top_extractor,
+            rpc_gateway,
+        }
+    }
+
+    fn method_impl(
+        &self,
+        request: DirectRequest,
+    ) -> Result<(Balances, bool, DirectRequestStatus), String> {
+        debug!("entering get_balance RPC");
+
+        let verified_trusted_operation =
+            self.top_extractor.get_verified_trusted_operation(request)?;
+
+        let trusted_getter_signed = match verified_trusted_operation {
+            TrustedOperation::get(getter) => match getter {
+                Getter::trusted(tgs) => Ok(tgs),
+                _ => Err(RpcCallStatus::operation_type_mismatch.to_string()),
+            },
+            _ => Err(RpcCallStatus::operation_type_mismatch.to_string()),
+        }?;
+
+        let main_account = trusted_getter_signed.getter.main_account().clone();
+        let proxy_account = trusted_getter_signed.getter.proxy_account().clone();
+
+        let _authorization_result = match self
+            .rpc_gateway
+            .authorize_user(main_account.clone(), proxy_account)
+        {
+            Ok(()) => Ok(()),
+            Err(e) => Err(format!("Authorization error: {}", e)),
+        }?;
+
+        let asset_id = match trusted_getter_signed.getter {
+            TrustedGetter::get_balance(_, c, _) => Ok(c),
+            _ => Err(RpcCallStatus::operation_type_mismatch.to_string()),
+        }?;
```

Comment:
Gets the currency id, and implicitly makes sure we have the correct trusted operation